alm/src/utils/contract.ts

```diff
@@ -19,33 +33,58 @@ export const getRequiredBlockConfirmations = async (contract: Contract, blockNum
   return parseInt(blockConfirmations)
 }
 
-export const getValidatorAddress = (contract: Contract) => contract.methods.validatorContract().call()
+export const getValidatorAddress = async (contract: Contract, snapshotProvider: SnapshotProvider) => {
```

Comment:
The validator contract address could be changed later so data in the snapshot could be outdated.   There is no big advantage if we get them form the snapshot.

Reply:
In which scenario do you think the validator contract address could be changed? Validator contracts are usually a proxy too and the bridge contract does not provide a method for such functionality

Reply:
Imagine that the bridge has lost the control over the validators contract e.g. the validators contract upgradability admin is not the same account as the bridge upgradability admin AND the private key of the validators contract upgradability admin was lost. So, we will need to stop the bridge and deploy another validator contract with the same parameters as an old one except the validators contract upgradability admin account.

Reply:
Thanks for the example!
In that case, as you suggested it makes sense to remove the validator contract address.

If the validator contract address is changed, right now there is no way to get the historic values of the validator contracts addresses. So if a user introduces a transaction that was previous to the change to a new validator contract, we won't be able to get the correct parameters of validator list and required signatures at that moment since we won't be able to get the validator contract of that moment and in the new contract, validators could be different.

Does it make sense to add a new event like `ValidatorContractChanged` when the bridge is initialized? it could be emitted also when the validator contract changes in some upgrade method invoked. Then in the future, ALM could use those events.

Reply:
> So if a user introduces a transaction that was previous to the change to a new validator contract, we won't be able to get the correct parameters of validator list and required signatures at that moment since we won't be able to get the validator contract of that moment and in the new contract, validators could be different.

> Does it make sense to add a new event like ValidatorContractChanged when the bridge is initialized? it could be emitted also when the validator contract changes in some upgrade method invoked. Then in the future, ALM could use those events.

OK. Let's introduce such event in the contracts. But we will not handle this situation in ALM so far since it is too rare.

Reply:
Removed the validator address from the snapshot https://github.com/poanetwork/tokenbridge/pull/382/commits/33c06eedd3637fcfb2c421d4e7ef1ae9e4b97d07
And also created an issue for the event in the contracts https://github.com/poanetwork/tokenbridge-contracts/issues/445